File: benchmarks/binary_trees/binary_trees.py
# binary_trees — allocation-churn benchmark (mirrors binary_trees.c / .rl).
# maxDepth=14; nodes are heap objects reclaimed by CPython's refcount/GC.

import sys


class Tree:
    __slots__ = ("left", "right")

    def __init__(self, left, right):
        self.left = left
        self.right = right


def make(depth):
    if depth == 0:
        return Tree(None, None)
    return Tree(make(depth - 1), make(depth - 1))


def check(t):
    if t is None:
        return 0
    return 1 + check(t.left) + check(t.right)


def main():
    min_depth = 4
    max_depth = 14
    total = 0

    # stretch tree
    stretch_depth = max_depth + 1
    stretch = make(stretch_depth)
    total += check(stretch)
    del stretch

    # long-lived tree
    long_lived = make(max_depth)

    # iteration loop
    depth = min_depth
    while depth <= max_depth:
        exp = max_depth - depth + min_depth
        iterations = 1 << exp
        for _ in range(iterations):
            t = make(depth)
            total += check(t)
        depth += 2

    # long-lived tree check
    total += check(long_lived)

    print(total)


if __name__ == "__main__":
    sys.setrecursionlimit(10000)
    main()

File: benchmarks/binary_trees/BinaryTrees.java
// binary_trees — allocation-churn benchmark (mirrors binary_trees.c / .rl).
// maxDepth=14; nodes are heap objects reclaimed by the JVM GC.
public class BinaryTrees {
    static final class Tree {
        Tree left, right;
    }

    static Tree make(int depth) {
        Tree t = new Tree();
        if (depth != 0) {
            t.left = make(depth - 1);
            t.right = make(depth - 1);
        }
        return t;
    }

    static long check(Tree t) {
        if (t == null) return 0;
        return 1 + check(t.left) + check(t.right);
    }

    public static void main(String[] args) {
        int minDepth = 4;
        int maxDepth = 14;
        long total = 0;

        // stretch tree
        int stretchDepth = maxDepth + 1;
        Tree stretch = make(stretchDepth);
        total += check(stretch);
        stretch = null;

        // long-lived tree
        Tree longLived = make(maxDepth);

        // iteration loop
        for (int depth = minDepth; depth <= maxDepth; depth += 2) {
            int exp = maxDepth - depth + minDepth;
            long iterations = 1L << exp;
            for (long i = 0; i < iterations; i++) {
                Tree t = make(depth);
                total += check(t);
            }
        }

        // long-lived tree check
        total += check(longLived);

        System.out.println(total);
    }
}
